keanu-project/src/main/java/io/improbable/keanu/tensor/INDArrayShim.java

```diff
@@ -24,76 +24,116 @@
+        if (a.shape().length < b.shape().length) {
+            return broadcastPlus(b, a);
+        }
+        return execBroadcast(a, b,
+            new BroadcastAddOp(a, b, a.dup(), broadcastDimensions)
         );
     }
 
-    public static void subi(INDArray left, INDArray right, INDArray result) {
+    public static INDArray subi(INDArray left, INDArray right) {
         if (Arrays.equals(left.shape(), right.shape())) {
-            left.subi(right);
+            return left.subi(right);
         } else {
-            broadcastMinus(left, right, result);
+            return broadcastMinus(left, right);
         }
     }
 
-    private static void broadcastMinus(INDArray a, INDArray b, INDArray result) {
+    private static INDArray broadcastMinus(INDArray a, INDArray b) {
         int[] broadcastDimensions = Shape.getBroadcastDimensions(a.shape(), b.shape());
-        execBroadcast(a, b,
-            new BroadcastSubOp(a, b, result, broadcastDimensions)
-        );
+        if (a.shape().length < b.shape().length) {
+            return broadcastPlus(a.neg(), b);
+        } else {
+            return execBroadcast(a, b,
+                new BroadcastSubOp(a, b, a.dup(), broadcastDimensions)
+            );
+        }
     }
 
-    private static void execBroadcast(INDArray a, INDArray b, BroadcastOp op) {
+    private static INDArray execBroadcast(INDArray a, INDArray b, BroadcastOp op) {
         int[] executeAlong = getBroadcastAlongDimensions(a.shape(), b.shape());
-        Nd4j.getExecutioner().exec(op, executeAlong);
+        return Nd4j.getExecutioner().exec(op, executeAlong);
+    }
+
+    private static int[] getBroadcastDimensions(int[] shapeA, int[] shapeB) {
+        int maxRank = Math.max(shapeA.length, shapeB.length);
+
+        if (shapeA.length < shapeB.length) {
+            shapeA = TensorShape.shapeToDesiredRankByPrependingOnes(shapeA, shapeB.length);
+        } else {
+            shapeB = TensorShape.shapeToDesiredRankByPrependingOnes(shapeB, shapeA.length);
+        }
+
+        List<Integer> along = new ArrayList<>();
+
+        for (int i = maxRank - 1; i >= 0; i--) {
+            if (shapeA[i] != shapeB[i]){
+                along.add(i);
+            }
+        }
+        return Ints.toArray(along);
     }
 
     private static int[] getBroadcastAlongDimensions(int[] shapeA, int[] shapeB) {
-        int minRank = Math.min(shapeA.length, shapeB.length);
+        int maxRank = Math.max(shapeA.length, shapeB.length);
+
+        if (shapeA.length < shapeB.length) {
+            shapeA = TensorShape.shapeToDesiredRankByPrependingOnes(shapeA, shapeB.length);
+        } else {
+            shapeB = TensorShape.shapeToDesiredRankByPrependingOnes(shapeB, shapeA.length);
+        }
+
         List<Integer> along = new ArrayList<>();
-        for (int i = 0; i < minRank; i++) {
-            if (shapeA[i] == shapeB[i]) {
+
+        for (int i = maxRank - 1; i >= 0; i--) {
+            if (shapeA[i] == shapeB[i]){
```

Comment:
Two functions that are very similar apart from a single operation - could we collapse the code down a bit?

Reply:
I'd normally agree - but the small difference between the functions is so crucial I'd rather it was overly explicit what was going on. I tried a boolean flag whether it's `along` or `matching` but it just makes something very simple a bit confusing